Fix Archetypes reference migration

rpatterson opened the pull request.

rpatterson committed:
Fix Folder migration dup indexing of children unindexing AT refs

Fixes #22

rpatterson committed:
Fix new UID's not being indexed for non-AT content

mister-roboto commented:
@rpatterson thanks for creating this Pull Request and help improve Plone!

To ensure that these changes do not break other parts of Plone, the Plone test suite matrix needs to pass.

Whenever you feel that the pull request is ready to be tested, either start all jenkins jobs pull requests by yourself, or simply add a comment in this pull request stating:

```
@jenkins-plone-org please run jobs
```

With this simple comment all the jobs will be started automatically.

Happy hacking!

rpatterson commented:
Required for https://github.com/plone/plone.app.contenttypes/issues/510
@jenkins-plone-org please run jobs

rpatterson commented:
Fixes #22

rpatterson committed:
Fix AT reference migration interfering with ReferenceField migration.

Fixes #22

rpatterson commented:
@jenkins-plone-org please run jobs

rpatterson committed:
Attempt to fix plone.app.blob Jenkins errors, #23

rpatterson commented:
@jenkins-plone-org please run jobs

pbauer commented:
I cannot really say if that is ok. I'm worried that there might be a side-effect to not stting the childen into `self._objects`. Skipping reindexing should be ok thought. 
If you are sure that migrations still work with this change I have no objections. You should add the same change to the master-branch as well though.

rpatterson committed:
Fix flake8 errors

rpatterson committed:
Fix rename indexing, update copied code from OFS.CopySupport

rpatterson committed:
Fix flake8 errors

rpatterson committed:
Fix CMFEditions being purged when old object is deleted

rpatterson committed:
Fix replacing automatically created content, breaks on KeyError

rpatterson committed:
Remove folder children optimization that breaks common usage:

There are too many things that can depend on catalog indexes being accurate
while creating the new folder for this optimization to be the default.  For
example, if any indexed value requires using the catalog to filter or
aggregate some value from children, then calling `getObject()` on those
catalog brains will raise an exception.  There are many potential use cases
that would run into this, but in my case it was an automatic workflow
transition that caused reindexing workflow variables.

rpatterson committed:
Cleanup flake8 errors

gforcada commented on Products/contentmigration/basemigrator/migrator.py:
```diff
@@ -574,23 +573,13 @@ def migrate_children(self):
-                    self.new._setOb(id, obj)
-                else:
-                    raise
+            if id in self.new.objectIds():
```
please avoid using python builtins, `object_id` would be as fine as well :wink:

rpatterson committed:
Fix expired and restricted content not being migrated

rpatterson committed:
Fix workflow variables being out-of-date in the catalog

Needed after removing the indexing optimizations.

rpatterson committed:
Fix auto-created child content remaining in catalog after removal:

Remove another optimization that breaks in common use cases.

rpatterson committed:
Add support for migrating marker interfaces.

rpatterson committed:
Add missing ChangeLog entries

pbauer commented:
I adapted these changes to the master-branch with https://github.com/plone/Products.contentmigration/pull/24 and the test `test_default_pages_are_kept_during_migration` is failing. 

@rpatterson The reason is your reimplementation of `migrate_children` which no longer stores the objects in `self.subobjs` and sets them again with `self.new._setObject()`. 

Can you comment why you chose to use `utils.unrestricted_move` instead? That seems to be less efficient since it triggers a lot of subscribers for the subobjects. And also for some reason the default_page is lost.

There are a lot of things here that I like but that change looks wrong to me.

dest81 commented:
@pbauer 
I am testing this changes now. I found only two bugs:
 - as you mentioned before default page is lost
 - children's order in the new folder is wrong 

Here is the commit to fix this two bugs https://github.com/dest81/Products.contentmigration/commit/4b316ac2fb97c5965ecea0ff393a06bcf73fc67e

I can send PR to apply_rpatterson_changes branch if you agree with it.

Let you know if I find more bugs

pbauer commented:
@dest81 yes, please create a pull-request

pbauer commented:
I finally looked at https://github.com/plone/Products.contentmigration/pull/23/commits/aa89e57243824b734204b4f64536e12c19f47195 (especially the commit-message) more closely and I now think that the change is ok and can be merged once the fixes by @dest81 are included.

But: For all of my use-cases the old optimized version also worked and I have not compared the performance of the two implementations.

pbauer commented:
@jenkins-plone-org please run jobs

jensens commented:
There was no activity on this for a while. I close this request for now. If you plan to work further on this please reopen and go on.

rpatterson commented:
> @rpatterson The reason is your reimplementation of `migrate_children` which no longer stores the objects in `self.subobjs` and sets them again with `self.new._setObject()`.
> 
> Can you comment why you chose to use `utils.unrestricted_move` instead? That seems to be less efficient since it triggers a lot of subscribers for the subobjects. And also for some reason the default_page is lost.

My apologies for not capturing the why for this and other changes in my commits, or at the very least in this PR's description.  I'm much better about that these days, not that it helps here.

At any rate, I dimly recall there being a cascade of problems in real-world, client data that arose from these optimizations.  IOW, there were real consequences to *not* running all those subscribers that broke things in real-world data.  But that's all I recall at this point so long after I made the changes.  Again, sorry.
